source/binding/OperatorExpressions.cpp

```diff
@@ -1498,6 +1501,145 @@ void ReplicationExpression::serializeTo(ASTSerializer& serializer) const {
     serializer.write("concat", concat());
 }
 
+Expression& StreamingConcatenationExpression::fromSyntax(
+    Compilation& compilation, const StreamingConcatenationExpressionSyntax& syntax,
+    const BindContext& context, const Type* assignmentTarget) {
+    const bool isRightToLeft = syntax.operatorToken.kind == TokenKind::LeftShift;
+    size_t sliceSize = 0;
+
+    auto badResult = [&]() {
+        return compilation.emplace<StreamingConcatenationExpression>(
+            compilation.getErrorType(), sliceSize, span<const Expression*>(), syntax.sourceRange());
+    };
+
+    if (!(context.flags & BindFlags::StreamingAllowed)) {
+        context.addDiag(diag::BadStreamContext, syntax.operatorToken.location());
+        return badExpr(compilation, badResult());
+    }
+
+    if (syntax.sliceSize) {
+        if (isRightToLeft) {
+            const auto& sliceExpr =
+                bind(*syntax.sliceSize, context, BindFlags::AllowDataType | BindFlags::Constant);
+            if (sliceExpr.bad())
+                return badExpr(compilation, badResult());
+            if (sliceExpr.kind == ExpressionKind::DataType) {
+                if (!sliceExpr.type->isFixedSize()) {
+                    auto& diag = context.addDiag(diag::BadStreamSlice, sliceExpr.sourceRange);
+                    if (sliceExpr.type->location)
+                        diag.addNote(diag::NoteDeclarationHere, sliceExpr.type->location);
+                    return badExpr(compilation, badResult());
+                }
+                sliceSize = sliceExpr.type->bitstreamWidth();
+            }
+            else {
+                optional<int32_t> count = context.evalInteger(sliceExpr);
+                if (!count)
+                    return badExpr(compilation, badResult());
+                if (*count <= 0) {
+                    context.requireGtZero(count, sliceExpr.sourceRange);
+                    return badExpr(compilation, badResult());
+                }
+                sliceSize = static_cast<bitwidth_t>(*count);
+            }
+        }
+        else
+            context.addDiag(diag::IgnoredSlice, syntax.sliceSize->sourceRange());
+    }
+    else if (isRightToLeft)
+        sliceSize = 1;
+
+    SmallVectorSized<Expression*, 8> buffer;
+    for (const auto argSyntax : syntax.expressions) {
+        Expression* arg;
+        if (assignmentTarget &&
+            argSyntax->expression->kind == SyntaxKind::StreamingConcatenationExpression)
+            arg = &create(compilation, *argSyntax->expression, context, BindFlags::StreamingAllowed,
```

Comment:
You probably don't need to pass the assignmentTarget through here, you aren't using it for anything right?

Reply:
I think I need to improve comments in this place.
1. I add
```c++
    // The sole purpose of assignmentTarget here is for Type::isBitstreamType to exclude associative
    // arrays and classes for target/destination. Streaming concatenation is self-determined and its
    // size and type should not be affected by assignmentTarget.
```
We need a way to know whether this streaming concatenation is a source or target, so Type::isBitstreamType can know whether to exclude associative arrays. Ideally, the "Type* assignmentTarget" argument should be replaced with "bool isDestination". However, this function is called through Expression::create or Expression::selfDetermined, and changing its arguments will require the change of those functions too. So I play a trick to reuse "assignmentTarget" for the purpose of isDestination.

2. I replace the comment "Streaming operator has no meaningful type. Use void as a placeholder." with
```c++
    // Streaming concatenation has no explicit type. Use void to prevent unintentional problems when
    // its type is accidentally passed to context-determine expressions.
```
When a streaming concatenation appears as the target of an assignment, the rhs of the assignment should be self-determined. The size, signedness and type compatibility checking of rhs should not be affected by the streaming concatenation, especially when the rhs is context-determined.

Reply:
I see, thanks, this helped me understand.